src/Illuminate/Support/Arr.php

```diff
@@ -517,24 +517,20 @@ public static function set(&$array, $key, $value)
             return $array = $value;
         }
 
-        $keys = explode('.', $key);
-
-        while (count($keys) > 1) {
-            $key = array_shift($keys);
-
-            // If the key doesn't exist at this depth, we will just create an empty array
-            // to hold the next value, allowing us to create the arrays to hold final
-            // values at the correct depth. Then we'll keep digging into the array.
-            if (! isset($array[$key]) || ! is_array($array[$key])) {
-                $array[$key] = [];
+        $segments = explode('.', $key);
+
+        foreach ($segments as $segment) {
+            // If the key doesn't exist at this depth, we'll just create an
+            // empty array to hold the next value, allowing us to create the
+            // arrays to hold the final value at the correct depth.
+            if (isset($array) && ! is_array($array)) {
```

Comment:
Why does this test even need to be inside the loop? Just set the `$array = []` outside of the `foreach` - then remove this whole section.

Reply:
Because we need to check this for every level of the (possibly) nested array.

Reply:
Since it's passed by reference it only overwrites the current level it's at.